Update Node.js Tutorial From Prisma1 To Prisma

robin-macpherson opened the pull request:
## Description

Update the Node.js tutorial to use Prisma instead of Prisma1.
I went through the first few sections in detail and made some minor updates: a few typos or small grammar fixes, updated all links (`graph.cool` links and also `zeit > vercel`, for example).

###  Notes

I have also done my best to significantly re-write the explanation of the what/why behind Prisma (section 4) and will be seeking feedback on my approach soon when it's ready!

Section 5 on Subscriptions is being broken out into a bonus chapter due to not being supported at this time. We will show how to implement them in a modern Prisma project.

## Sub-Tasks

- [x] Run through sections 0-3 and make any small updates such as typos or grammar fixes
- [x] Update links
- [x] Re-write section 4 (adding a database) to have new Prisma information and explanation
- [x] Update section 4 code/implementation
- [x] Update section 5 (connecting server and db)
- [x] Update section 6 (authentication)
- [x] Break section 7 (subscriptions) into bonus chapter
- [x] Update section 8 (filtering & pagination)
- [x] Update summary

robin-macpherson commented on content/backend/graphql-js/4-adding-a-database.md:
```diff
@@ -7,189 +7,235 @@ answers: ["To increase performance of the server", "When using two APIs, the Gra
-  const allLinks = await prisma.links()
-  console.log(allLinks)
-}
+</Instruction>
```
@nikolasburk the reason that I decided to split up this from the original query is because I think the autocompletion is really so good and helpful and I wanted to sort of "train" them to use this autocompletion to help them understand how to write new queries/mutations themselves. But let me know if you think it's a bit dis-jointed or redundant to split it up into separate steps like this 🙂

nikolasburk commented on content/backend/graphql-js/0-introduction.md:
```diff
@@ -16,20 +16,20 @@ GraphQL is the rising star of backend technologies. It replaces REST as an API d
-  * ... provides an editor where you can write queries, mutations & subscriptions, with auto-completion(!) and syntax highlighting.
-  * ... lets you easily share your API operations.
+* [`graphql-yoga`](https://github.com/prisma/graphql-yoga): A fully-featured GraphQL server with a focus on easy setup, performance & great developer experience. It is built on top of [Express](https://expressjs.com/), [`apollo-server`](https://github.com/apollographql/apollo-server), [`graphql-js`](https://github.com/graphql/graphql-js) and more.
+* [Prisma](https://www.prisma.io/): Replaces traditional ORMs. Use the Prisma client to implement your GraphQL resolvers and simplify database access.
```
can you adjust "the Prisma client" to just "Prisma Client", incl removing the article? (here and wherever else you come across it 🙂 )

robin-macpherson replied:
Done! ✅🙂

nikolasburk commented on content/backend/graphql-js/0-introduction.md:
```diff
@@ -16,20 +16,20 @@ GraphQL is the rising star of backend technologies. It replaces REST as an API d
+You'll start by learning the basics of how a GraphQL server works, simply by defining a [_GraphQL schema_](https://www.prisma.io/blog/graphql-server-basics-the-schema-ac5e2950214e) for the server and writing corresponding _resolver functions_. In the beginning, these resolvers will only work with data that's stored in-memory - so nothing will persist beyond the runtime of the server.
 
-Because nobody wants a server that's not able to store and persist data, you're going to add a database layer to it. The database layer is powered by [Prisma](https://www.prisma.io/) and will be connected to your GraphQL server via the [Prisma client](https://www.prisma.io/docs/prisma-client). 
+Because nobody wants a server that's not able to store and persist data, you're going to add a database layer to it. The database layer is powered by [Prisma](https://www.prisma.io/) and will be connected to your GraphQL server via the [Prisma client](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/api#overview). 
```
can you adjust "the Prisma client" to just "Prisma Client", incl removing the article? (here and wherever else you come across it 🙂 )

robin-macpherson replied:
Done! ✅🙂

nikolasburk commented on content/backend/graphql-js/1-getting-started.md:
```diff
@@ -73,7 +73,7 @@ Here's a list of its features:
 - Query performance tracing
 - Accepts both `application/json` and `application/graphql` content-types
-- Runs everywhere: Can be deployed via `now`, `up`, AWS Lambda, Heroku etc.
+- Runs everywhere: Can be deployed via `vercel`, `up`, AWS Lambda, Heroku etc.
```
nice catch 😄

robin-macpherson replied:
heheh 😉

nikolasburk commented on content/backend/graphql-js/4-adding-a-database.md:
```diff
@@ -7,189 +7,235 @@ answers: ["To increase performance of the server", "When using two APIs, the Gra
 
-- Access the database directly (by writing SQL or using another NoSQL database API)
-- Use an [ORM](https://en.wikipedia.org/wiki/Object-relational_mapping) that provides an abstraction for your database and lets you access it directly from your programming language
+### Why Prisma?
```
I really like the new version of this section, feels a lot more focussed and fits the flow of the tutorial 👍

robin-macpherson replied:
That's fantastic, such a relief to read that 😀🙌🏼

nikolasburk commented on content/backend/graphql-js/4-adding-a-database.md:
```diff
@@ -7,189 +7,235 @@ answers: ["To increase performance of the server", "When using two APIs, the Gra
 </Instruction>
 
-[`prisma.yml`](https://www.prisma.io/docs/-5cy7/) is the main configuration file for your Prisma setup. `datamodel.prisma` on the other hand contains the definition of your [datamodel](https://www.prisma.io/docs/-knul). The Prisma datamodel defines your application's _models_. Each model will be mapped to a table in the underlying database.
+Remember the GraphQL schema that you've been working with until now? Well, Prisma has a schema, too! You can think of the `prisma.schema` file as a *database schema*. It has three components:
```
Nice! 👍

nikolasburk commented on content/backend/graphql-js/4-adding-a-database.md:
```diff
@@ -7,189 +7,235 @@ answers: ["To increase performance of the server", "When using two APIs, the Gra
+1. **Generator**: Indicates that you want to genenerate Prisma Client.
+1. **Data model**: Defines your application *models*. Each model will be mapped to a table in the underlying database.
+
+Prisma's unique data model bridges the gap to help you reason about your data in a way that maps very well to the underlying database, while still providing an abstraction that allows you to be productive with type safety and superb auto-completion.
```
While I agree with the sentiment, I'd remove the "superb", feels a little too marketing-y here 🙈

robin-macpherson replied:
hahah I felt conflicted about it as I was writing it 🙈 it really is bloody superb, but I understand 🙂

nikolasburk commented on content/backend/graphql-js/4-adding-a-database.md:
````diff
@@ -7,189 +7,235 @@ answers: ["To increase performance of the server", "When using two APIs, the Gra
+npx prisma migrate save --experimental
 ```
 
+You will get a prompt asking if you would like to create a new database. Select "Yes", and type "Init DB" for the `Name of migration`.
````
I believe the `<Instruction>` component is too limited for this, it can only have: 

1. normal text
1. code block

In this case you have:

1. normal text
1. code block
1. normal text

Which I think breaks it iirc, best factor the prompt into itsown `<Instruction>` block.

robin-macpherson replied:
ah! Good catch, I wasn't aware of the limitation. Just refactored it and will keep that in mind moving forward 🙂

nikolasburk commented on content/backend/graphql-js/4-adding-a-database.md:
```diff
@@ -7,189 +7,235 @@ answers: ["To increase performance of the server", "When using two APIs, the Gra
+Take a look at the `prisma` directory in your project's file system now. You'll see that there is now a `/migrations` directory that was created for you when running `npx prisma migrate save --experimental`. 
 
-Here's a quick explanation of each property you see in that file:
+For now, the important thing to understand is that we have told Prisma with our data model, "I want create a `Link` table to store Link data, and here's what that data will look like. Prisma then generates the necessary migration and packages it into a dedicated directory with its own `README.md` file containing detailed information about the specific migration. This is then put inside that `pris ...
```
Nice 👌

robin-macpherson commented:
@nikolasburk I've addressed all comments up to completing section 4 and will let you know when section 5 is ready for review (or if I have open questions) shortly 🚀

nikolasburk commented:
That's awesome, looking forward to the next ones 🙌

robin-macpherson commented on content/backend/graphql-js/0-introduction.md:
```diff
@@ -16,20 +16,20 @@ GraphQL is the rising star of backend technologies. It replaces REST as an API d
+* [GraphQL Playground](https://github.com/prisma/graphql-playground): A "GraphQL IDE" that allows you to interactively explore the functionality of a GraphQL API by sending queries and mutations to it. It's somewhat similar to [Postman](https://www.getpostman.com/) which offers comparable functionality for REST APIs. Among other things, a GraphQL Playground:
+  * Auto-generates comprehensive documentation for all available API operations.
+  * Provides an editor where you can write queries, mutations & subscriptions, with auto-completion(!) and syntax highlighting.
+  * Lets you easily share your API operations.
```
@nikolasburk as I'm reading back through the sections thus far, I'm wondering if we should include any information about the Prisma technologies we'll be using here? For example, the fact that we mention GraphQL Playground makes me feel like Prisma Studio could totally be worth mentioning. 

Maybe I can add sub-bullet points underneath *Prisma* for Prisma Client, Prisma Migrate, and Prisma Studio? Let me know what you think 🙂

nikolasburk replied:
Good idea, let's do it! 💪

robin-macpherson commented on content/backend/graphql-js/1-getting-started.md:
```diff
@@ -219,6 +219,8 @@ In this case, we have three root fields: `users` and `user` on `Query` as well a
+
 - `typeDefs`: These are the type definitions from your application schema.
 - `resolvers`: This is a JavaScript object that mirrors the `Query`, `Mutation` and `Subscription` types and their fields from your application schema. Each field in the application schema is represented by a function with the same name in that object.
 - `context`: This is an object that gets passed through the resolver chain and every resolver can read from or write to.
```
@nikolasburk as I was reading back through the original section here, I feel like there is something missing here? We talk about Scalar Types and Object Types and then 💥boom, 3 bullet points explaining `typeDefs`, `resolvers`, and `context`. It feels quite abrupt. Feels like there should be a sentence right before the three bullet points to lead into it.

nikolasburk replied:
Hmm I agree, this seems like an error actually... might have been a leftover from a previous iteration 🙈  I don't think the bullets are properly placed in this section at all, just removing them I think would probably resolve this?

robin-macpherson replied:
Yep it seemed like an error to me, too. I was going to just try to fix it but I simply couldn't see how the bullets made sense here 😅 I agree that simply removing them resolves is since the following paragraph leads on perfectly 😄 Done ✅

robin-macpherson commented on content/backend/graphql-js/1-getting-started.md:
```diff
@@ -256,13 +258,13 @@ mutation {
   - For the `createUser(name: String!)` field, the return type `User!` means this operation always returns a `User` object.
 
+// TODO (robin-macpherson): question #2 for Nikolas
 As you provide this information, the `Prisma` instance will get full access to your database service and can be used to resolve incoming request later on.
```
@nikolasburk it feels random that we mentioned Prisma suddenly here. Was this intentional?

nikolasburk replied:
Huh,... yeah that seems weird! Maybe a similar issue as the other one where this might be just a leftover from a previous iteration 🤔 if this doesn't make sense in this context feel free to remove it.

robin-macpherson replied:
Ok will do!

robin-macpherson commented on content/backend/graphql-js/4-adding-a-database.md:
```diff
@@ -7,189 +7,238 @@ answers: ["To increase performance of the server", "When using two APIs, the Gra
+Prisma's unique data model bridges the gap to help you reason about your data in a way that maps very well to the underlying database, while still providing an abstraction that allows you to be productive with type safety and auto-completion.
 
-So far, the datamodel for your Hacker News app only contains one data type: `Link`. Because Prisma uses [GraphQL SDL](https://www.prisma.io/blog/graphql-sdl-schema-definition-language-6755bcb9ce51) for model definitions, you can basically copy the existing `Link` definition from `schema.graphql` into `datamodel.prisma`.
+Let's see it in action with your project!
```
@nikolasburk I wonder if it is worth at some point mentioning the Prisma VS Code extension? With Journaly, I was recently very impressed with certain things such as how it even has the ability to automatically add in naming both sides of a relationship, etc. and it's pretty darn powerful! Even just the auto-formatting is very nice.

I could see why we would not mention it. It's just a thought if we might want to mention that certain editors have a great extension you can install for even more productivity benefits, syntax highlighting, etc.

nikolasburk replied:
Yes, absolutely! Let's definitely mention the VS Code extension (and the equivalents for Jetbrains IDEs: https://plugins.jetbrains.com/plugin/14240-prisma)

robin-macpherson replied:
Ok got it!! 🙌🏼

robin-macpherson commented:
@nikolasburk first draft of `section-5` is up!! 😄🎸

robin-macpherson commented on content/backend/graphql-js/6-authentication.md:
```diff
@@ -11,52 +11,105 @@ In this section, you're going to implement signup and login functionality that a
 </Instruction>
 
-This now updated the Prisma API. The Prisma client is generated automatically after running `prisma deploy`, exposing all the CRUD methods for the newly added User model.
+// TODO @nikolasburk this command actually failed here. Temporarily I had to just delete all migrations in order to move on but we should make sure that everyone isn't going to experience this issue.
```
@nikolasburk just to point this out for you ⬆️

nikolasburk replied:
Hmm, weird! What was the error message you saw?

robin-macpherson commented on content/backend/graphql-js/8-filtering-pagination-and-sorting.md:
```diff
@@ -76,30 +81,31 @@ query {
-- The _limit_ is called `first`, meaning you're grabbing the _first_ x elements after a provided start index. Note that you also have a `last` argument available which correspondingly returns the _last_ x elements.
-- The _start index_ is called `skip`, since you're skipping that many elements in the list before collecting the items to be returned. If `skip` is not provided, it's `0` by default. The pagination then always starts from the beginning of the list (or the end in case you're using `last`).
+// TODO (robin-macpherson): @nikolasburk it seems we still need to update this over in the `grraphql-js` project as well
+- The _limit_ is called `take`, meaning you're "taking" x elements after a provided start index.
```
@nikolasburk this was a little weird because the Prisma docs say that this is now called `take` but I was coding along with Journaly side-by-side testing everything out and it seemed to still be `first` and I saw that you still had it as `first` in the completed project code. Is there some discrepancy here?

nikolasburk replied:
Ah sorry, yes we had a breaking change in the pagination API in [2.0.0-beta.7](https://github.com/prisma/prisma/releases/tag/2.0.0-beta.7):

- Removed `first`, `last`, `before`, `after` arguments.
- Added `cursor` and `take` arguments.
- `skip` argument unchanged.

The `take` argument replaces `first` and `last`.

Where did you see outdated Prisma docs that were using the old pagination API, we should make sure that this gets updated!

robin-macpherson commented on content/backend/graphql-js/6-authentication.md:
```diff
@@ -43,17 +42,19 @@ model User {
-Notice how you're adding a new _relation field_ called `postedBy` to the `Link` model that points to a `User` instance. The `User` model then has a `links` field that's a list of `Link`s. To do this, we need to also define the relation by annotating the `postedBy` field with [the `@relation` attribute](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations#the-relation- ...
+Notice how you're adding a new _relation field_ called `postedBy` to the `Link` model that points to a `User` instance. The `User` model then has a `links` field that's a list of `Link`s. 
+
+o do this, we need to also define the relation by annotating the `postedBy` field with [the `@relation` attribute](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations#the-relation-attribute). This is required for every relation field in your Prisma schema, and all you're doing is defining what the foreign key of the related table will be. So in this case, we're addin ...
```
Missing a "T" in "To do this..."

robin-macpherson commented on content/backend/graphql-js/6-authentication.md:
```diff
@@ -43,17 +42,19 @@ model User {
-Notice how you're adding a new _relation field_ called `postedBy` to the `Link` model that points to a `User` instance. The `User` model then has a `links` field that's a list of `Link`s. To do this, we need to also define the relation by annotating the `postedBy` field with [the `@relation` attribute](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations#the-relation- ...
+Notice how you're adding a new _relation field_ called `postedBy` to the `Link` model that points to a `User` instance. The `User` model then has a `links` field that's a list of `Link`s. 
+
+o do this, we need to also define the relation by annotating the `postedBy` field with [the `@relation` attribute](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations#the-relation-attribute). This is required for every relation field in your Prisma schema, and all you're doing is defining what the foreign key of the related table will be. So in this case, we're addin ...
```
```suggestion
To do this, we need to also define the relation by annotating the `postedBy` field with [the `@relation` attribute](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations#the-relation-attribute). This is required for every relation field in your Prisma schema, and all you're doing is defining what the foreign key of the related table will be. So in this case, we're adding an extra field to store the `id` of the `User` who posts a `Link`, and then telling Prisma that `postedById` will be equal to the `id` field in the `User` table.
```

robin-macpherson replied:
(Suggestion just adds a missing "T" at the start of the line)
